### PR TITLE
First version of WikidataIdsDecorator

### PR DESCRIPTION
**Pull request by chrismytton**
This adds the implementation that was originally created in https://github.com/everypolitician-scrapers/saint-lucia-assembly-wikipedia/pull/1, along with documentation and tests.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,28 +1,35 @@
 # WikidataIdsDecorator
 
-Welcome to your new gem! In this directory, you'll find the files you need to be able to package up your Ruby library into a gem. Put your Ruby code in the file `lib/wikidata_ids_decorator`. To experiment with that code, run `bin/console` for an interactive prompt.
-
-TODO: Delete this and the text above, and describe your gem
+[Scraped](https://github.com/everypolitician/scraped) decorator which adds a `data-` attribute to Wikipedia links containing the Wikidata ID for the linked item, if relevant.
```
Comment (tmtmtmtm) [inline README.md]: This sentence is a little hard to grok, but I don't think it's accurate. AFAICS it actually adds a `wikidata` attribute, not a `data-` one. The "if relevant" is also a little hard to follow here. I think it's worth being a lot more explicit that the code actually goes and looks these up. Maybe something like?:

> When scraping a page from Wikipedia, this will attempt to look up the associated Wikidata item for each other Wikipedia page linked from within the page content, and adds a `wikidata="Qxxx"` attribute to each successfully resolved link.

File: lib/wikidata_ids_decorator.rb [hunk 2]
```diff
@@ -1,5 +1,25 @@
...
+    end
+
+    private
+
+    def wikidata_ids(links)
+      WikiData.ids_from_pages(language, links)
+    end
+
+    def language
+      'en'
```
Comment (tmtmtmtm) [inline lib/wikidata_ids_decorator.rb]: I don't think we can hardcode this to 'en' here. This will need to derive this from the `url` of the page being scraped. (A more advanced version will need to cope with interwiki links, but that can come later)

File: test/wikidata_ids_decorator_test.rb [hunk 3]
```diff
@@ -4,4 +4,18 @@
...
+    it 'adds a data-wikidata attribute to links' do
+      body = <<-HTML
+<div id="bodyContent">
+  <a href="https://en.wikipedia.org/wiki/Douglas_Adams" title="Douglas Adams">Douglas Adams</a>
+</div>
+      HTML
+      response = Scraped::Response.new(body: body, url: 'https://en.wikipedia.org/wiki/Douglas_Adams_(disambiguation)')
+      decorator = WikidataIdsDecorator::Links.new(response: response)
+      decorator.body.must_include '<a href="https://en.wikipedia.org/wiki/Douglas_Adams" title="Douglas Adams"' \
+        ' wikidata="Q42">Douglas Adams</a>'
```
Comment (tmtmtmtm) [inline test/wikidata_ids_decorator_test.rb]: That's an unfortunate place to have a line-break, as it makes it hard to tell whether there's something important happening there…

File: test/wikidata_ids_decorator_test.rb [hunk 4]
```diff
@@ -4,4 +4,18 @@
   it 'has a version number' do
     ::WikidataIdsDecorator::VERSION.wont_be_nil
   end
+
+  describe 'Wikipedia page with links' do
+    it 'adds a data-wikidata attribute to links' do
+      body = <<-HTML
+<div id="bodyContent">
+  <a href="https://en.wikipedia.org/wiki/Douglas_Adams" title="Douglas Adams">Douglas Adams</a>
+</div>
```
Comment (tmtmtmtm) [inline test/wikidata_ids_decorator_test.rb]: I think this falls into a recurring trap of making the test data function differently from what real data looks like — e.g. here we have absolute URLs, whereas wikipedia uses relative ones, and by not having any links outside `#bodyContent`, we aren't testing that those don't get included.  It's also not obvious here what will happen with edge cases like anchored links (e.g. when scraping https://en.wikipedia.org/wiki/Life,_the_Universe_and_Everything that it does the right thing with `<a href="/wiki/OCLC#Identifiers_and_linked_data" title="OCLC">`) or with links that are redirects (from the same page, `<a href="/wiki/Vogons" class="mw-redirect" title="Vogons">`)

I'm also not convinced we should be doing live lookups in the test, so if we're going to have to stub or VCR record the Wikidata lookup, we should probably do likewise for a real Wikipedia page too (and probably more than one, as we'll want to test that it works for more than en.wikipedia)

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]

File: test/wikidata_ids_decorator_test.rb [hunk 5]
```diff
@@ -1,21 +1,51 @@
...
+
+    it 'does not add a data-wikidata attribute to anchored links pointing to a place within the same page' do
+      decorator.body.must_include '<a href="#Plot_summary">'
+    end
+  end
+
+  describe 'Redirect links' do
+    let(:url) { 'https://en.wikipedia.org/wiki/Life,_the_Universe_and_Everything' }
+
+    it 'adds a data-wikidata attribute to redirect <links>  </links>' do
```
Comment (tmtmtmtm) [inline test/wikidata_ids_decorator_test.rb]: I don't really follow what the `<links>  </links>` is doing here. I'm also not sure that this really explains well what's being tested — surely this test is about making sure that it links to the _correct_ Wikidata item — i.e. after following the redirect?

File: test/wikidata_ids_decorator_test.rb [hunk 6]
```diff
@@ -1,21 +1,51 @@
...
   end
+end
+
+describe WikidataIdsDecorator::Links do
+  around { |test| VCR.use_cassette(File.basename(url), &test) }
+  let(:response) { Scraped::Request.new(url: url).response }
+  let(:decorator) { WikidataIdsDecorator::Links.new(response: response) }
 
   describe 'Wikipedia page with links' do
+    let(:url) { 'https://en.wikipedia.org/wiki/Douglas_Adams_(disambiguation)' }
```
Comment (tmtmtmtm) [inline test/wikidata_ids_decorator_test.rb]: I don't really follow why you're including this page, to test against, if all you're testing is that you get a Q42 link. If that link is also available from '/wiki/Life,_the_Universe_and_Everything', why not make it one of the tests against that? (which would also let you simplify things more as every section would be against the same `:url`)

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]
There are still a couple of things here that seem to be untested, and are quite important:

1. Links containing ':' within the body (e.g. the link to '/wiki/Special:BookSources/0-330-26738-8'
2. A page from a different language Wikipedia. (i.e. the current tests pass with language still hard-coded to `en`, when they should fail) [It might have been better to fix that bug in the original implementation as a PR against this PR, with `ensure_regression_tests` in action, rather than just committing it directly here. Even if you're not doing that, this is a significant enough change to deserve a commit of its own, not just a fixup.]

I note that currently running the tests also produces a `Can't find Wikidata IDs for: Technology in The Hitchhiker's Guide to the Galaxy, Infinite Improbability Drive, The Meaning of Liff in en` warnings. That should really be trapped and silenced (ditto the above suggestion for doing it as a PR to this PR, though I'm also open to doing it as a separate bug-report and new PR after this one gets merged)

**Comment by ondenman** [comment]
@tmtmtmtm I will open a separate PR for the language code. Do you mind if I remove 21a80c1 and force push? Or would the best thing be to undo that change in a fixup?

**Comment by ondenman** [comment]
I've removed 21a80c1 and force pushed.

**Comment by ondenman** [comment]
Checks are currently failing because of Rubocop. I'm fixing this in: https://github.com/everypolitician/wikidata_ids_decorator/pull/5

**Comment by ondenman** [comment]
I've opened an issue for the warning messages: https://github.com/everypolitician/wikidata_ids_decorator/issues/7

File: test/cassettes/Douglas_Adams__disambiguation_.yml [hunk 7]
```diff
@@ -0,0 +1,707 @@
+---
+http_interactions:
+- request:
+    method: get
+    uri: https://en.wikipedia.org/wiki/Douglas_Adams_(disambiguation)
```
Comment (tmtmtmtm) [inline test/cassettes/Douglas_Adams__disambiguation_.yml]: Is this cassette used?

File: test/wikidata_ids_decorator_test.rb [hunk 8]
```diff
@@ -1,7 +1,62 @@
...
+        decorator.body.must_include '<a href="/wiki/OCLC#Identifiers_and_linked_data" title="OCLC" wikidata="Q190593">'
+      end
+
+      it 'does not add a data-wikidata attribute to anchored links pointing to a place within the same page' do
+        decorator.body.must_include '<a href="#Plot_summary">'
+      end
+    end
+
+    describe 'Redirect links' do
+      it 'it follows the redirect and adds the correct wikidata item id' do
```
Comment (tmtmtmtm) [inline test/wikidata_ids_decorator_test.rb]: there's a doubled up 'it' here

File: test/wikidata_ids_decorator_test.rb [hunk 9]
```diff
@@ -1,7 +1,62 @@
...
+    end
+
+    describe 'Links outside of #bodyContent' do
+      it 'should leave these links alone' do
+        decorator.body.must_include '<a href="/wiki/Wikipedia:About" title="Wikipedia:About">About Wikipedia</a>'
+      end
+    end
+
+    describe 'URLs containing a colon' do
+      it 'it should leave links not pointing to an article alone' do
```
Comment (tmtmtmtm) [inline test/wikidata_ids_decorator_test.rb]: doubled 'it' again here, plus "should" style.

File: test/wikidata_ids_decorator_test.rb [hunk 10]
```diff
@@ -1,7 +1,62 @@
...
+    end
+
+    describe 'Redirect links' do
+      it 'it follows the redirect and adds the correct wikidata item id' do
+        decorator.body.must_include '<a href="/wiki/Vogons" class="mw-redirect" title="Vogons" wikidata="Q1076157">'
+      end
+    end
+
+    describe 'Links outside of #bodyContent' do
+      it 'should leave these links alone' do
```
Comment (tmtmtmtm) [inline test/wikidata_ids_decorator_test.rb]: This one introduces a mismatch of naming styles. All the other descriptions are asserting what _does_ happen. This one uses the "should" style. We're not quite as proscriptive on this as http://betterspecs.org/#expect, but we should be consistent.

**Review by tmtmtmtm — CHANGES_REQUESTED** [review]
A few small issues, but mostly looking good

**Review by tmtmtmtm — APPROVED** [review]
Woot!
